lib/b3fs/src/bucket/dir/state/mod.rs

```diff
@@ -0,0 +1,314 @@
+        file.write_all(&vec![0; disps_len * 4]).await?;
+        // 4 bytes for each entry
+        file.write_all(&vec![0; num_entries * 4]).await?;
+
+        let position_start_hashes = 8;
+        let position_start_bloom_filter = position_start_hashes + (num_entries * 32) as u64;
+        let position_start_phf_table = position_start_bloom_filter + num_entries as u64 + 4;
+        Ok(Self {
+            path,
+            file,
+            position_start_hashes,
+            position_start_bloom_filter,
+            position_start_phf_table,
+        })
+    }
+
+    /// Commits the HeaderFile by writing all necessary data and moving it to its final location.
+    ///
+    /// # Arguments
+    ///
+    /// * `bucket` - The Bucket to which this HeaderFile belongs.
+    /// * `root_hash` - The root hash of the directory.
+    /// * `hashes` - A vector of hashes for all entries.
+    /// * `bloom_filter` - The Bloom filter for quick membership tests.
+    /// * `hash_table` - The perfect hash function table.
+    ///
+    /// # Returns
+    ///
+    /// Returns a Result indicating success or a CommitError.
+    pub(crate) async fn commit(
+        mut self,
+        bucket: &Bucket,
+        root_hash: &[u8; 32],
+        hashes: Vec<[u8; 32]>,
+        bloom_filter: BloomFilter,
+        hash_table: HasherState,
+    ) -> Result<(), errors::CommitError> {
+        let mut file = self.file.into_inner().try_into_std().unwrap();
+        let bytes = hashes.iter().flatten().copied().collect::<Vec<u8>>();
+        tokio::task::spawn_blocking(move || async move {
+            // Write the hashes
+            file.write_at(&bytes, self.position_start_hashes)
+                .map_err(|e| io::Error::new(io::ErrorKind::Other, e))?;
+
+            // Write the bloom filter
+            file.write_at(
+                &bloom_filter.hashes().to_le_bytes(),
+                self.position_start_bloom_filter,
+            )
+            .map_err(|e| io::Error::new(io::ErrorKind::Other, e))?;
+            file.write_at(
+                bloom_filter.get_u8_array(),
+                self.position_start_bloom_filter + 4,
+            )
+            .map_err(|e| io::Error::new(io::ErrorKind::Other, e))?;
+
+            // Write the PHF table
+            let hash_table = hash_table;
+            let disp_bytes = bincode::serialize(&hash_table.disps)
+                .map_err(|e| errors::CommitError::Serialization(e.to_string()))?;
+            let map_bytes = bincode::serialize(&hash_table.map)
+                .map_err(|e| errors::CommitError::Serialization(e.to_string()))?;
+            file.write_at(&hash_table.key.to_le_bytes(), self.position_start_phf_table)?;
+            file.write_at(&disp_bytes, self.position_start_phf_table + 8)?;
+            file.write_at(
+                &map_bytes,
+                self.position_start_phf_table + 8 + disp_bytes.len() as u64,
+            )?;
+            Ok(()) as Result<(), errors::CommitError>
+        })
```

Comment:
is there a way to break this down into a series of async functions?

Reply:
The issue with split them up is that we end up having a bunch of async functions but are `spawn_blockinng` because you are dealing directly with the File OS directly, since there is no abstraction to write in a position on `tokio`. So i guess having multiple `spawn_blocking` will not be faster because the runtime will allocate all the async `spawn_blocking` in the same OS thread.

Reply:
I see.